### PR TITLE
Add logic to backdate snapshots for entities with provided start times

### PR DESCRIPTION
**Pull request by zdg2102**
Also add logging to `update_historical_snapshots` method

### CODE REVIEW

File: recidiviz/tests/persistence/database/database_test.py [hunk 1]
```diff
@@ -637,6 +638,187 @@ def testWritePeople_duplicatePeople_raisesError(self):
 
         session.close()
 
+    def testWritePerson_backdatedBooking_backdatesSnapshot(self):
+        name = 'some name'
```
Comment (terinpw) [inline recidiviz/tests/persistence/database/database_test.py]: nit - pull this out to a constant?

File: recidiviz/tests/persistence/database/database_test.py [hunk 2]
```diff
@@ -637,6 +638,187 @@ def testWritePeople_duplicatePeople_raisesError(self):
...
+        self.assertEqual(len(booking_snapshots), 2)
+        self.assertEqual(
+            booking_snapshots[0].valid_from, initial_scrape_time)
+        self.assertEqual(
+            booking_snapshots[1].valid_from, update_scrape_time)
+
+        assert_session.commit()
+        assert_session.close()
+
+    def testWritePerson_newPersonWithBackdatedBooking_backdatesPerson(self):
```
Comment (terinpw) [inline recidiviz/tests/persistence/database/database_test.py]: can we change the ending to `backdatesPersonSnapshot` just to be super explicit?

File: recidiviz/persistence/database/update_historical_snapshots.py [hunk 3]
```diff
@@ -161,9 +173,62 @@ def _fetch_most_recent_snapshots_for_entity_type(
...
+        earliest_admission_date = None
+
+        for booking in person.bookings:
+            admission_date = None
+            # Don't include case where admission_date_inferred is None
+            if booking.admission_date and \
+                    booking.admission_date_inferred is False:
+                admission_date = booking.admission_date
+                context_registry.snapshot_context(booking) \
+                    .provided_start_time = _date_to_datetime(admission_date)
```
[the remark this answers is not in the document]
Comment (zdg2102) [reply recidiviz/persistence/database/update_historical_snapshots.py]: `provided_start_time` gets passed to the `valid_from` column on the historical tables, which is a datetime

File: recidiviz/persistence/database/update_historical_snapshots.py [hunk 4]
```diff
@@ -161,9 +173,62 @@ def _fetch_most_recent_snapshots_for_entity_type(
...
+
+
+def _set_provided_start_time_for_booking_descendant(
+        entity: Any, parent_booking_admission_date: date,
+        context_registry: '_SnapshotContextRegistry') -> None:
+    """Sets |entity| provided start time on |context_registry| according to
+    type of |entity|
+    """
+
+    # TODO(1147): add type-specific logic
```
[the remark this answers is not in the document]
Comment (zdg2102) [reply recidiviz/persistence/database/update_historical_snapshots.py]: Yep, I'll include that before I close #1147

File: recidiviz/tests/persistence/database/database_test.py [hunk 5]
```diff
@@ -637,6 +638,187 @@ def testWritePeople_duplicatePeople_raisesError(self):
...
+        persisted_person = database.write_person(
+            arrange_session, person, IngestMetadata(
+                _REGION, _JURISDICTION_ID, person_scrape_time, {}))
+        arrange_session.commit()
+        person_id = persisted_person.person_id
+        arrange_session.close()
+
+        act_session = Session()
+        queried_person = database.read_people(
+            session=act_session, full_name=name, birthdate=birthdate)[0]
```
[the remark this answers is not in the document]
Comment (zdg2102) [reply recidiviz/tests/persistence/database/database_test.py]: Done

File: recidiviz/tests/persistence/database/database_test.py [hunk 6]
```diff
@@ -637,6 +638,187 @@ def testWritePeople_duplicatePeople_raisesError(self):
...
+            custody_status=CustodyStatus.IN_CUSTODY,
+            admission_date=booking_admission_date,
+            admission_date_inferred=False,
+            last_seen_time=booking_scrape_time)
+        queried_person.bookings = [booking]
+        updated_person = database.write_person(
+            act_session, queried_person, IngestMetadata(
+                _REGION, _JURISDICTION_ID, booking_scrape_time, {}))
+        act_session.commit()
+        booking_id = updated_person.bookings[0].booking_id
```
[the remark this answers is not in the document]
Comment (zdg2102) [reply recidiviz/tests/persistence/database/database_test.py]: Done throughout

File: recidiviz/tests/persistence/database/database_test.py (hunk 2 above)
Comment (zdg2102) [reply recidiviz/tests/persistence/database/database_test.py]: Unfortunately adding "snapshot" would put it over the line limit

File: recidiviz/tests/persistence/database/database_test.py (hunk 1 above)
Comment (zdg2102) [reply recidiviz/tests/persistence/database/database_test.py]: Done

File: recidiviz/tests/persistence/database/database_test.py [hunk 7]
```diff
@@ -637,6 +638,168 @@ def testWritePeople_duplicatePeople_raisesError(self):
...
+            arrange_session, person, IngestMetadata(
+                _REGION, _JURISDICTION_ID, initial_scrape_time, {}))
+        arrange_session.commit()
+        booking_id = persisted_person.bookings[0].booking_id
+        arrange_session.close()
+
+        act_session = Session()
+        queried_person = one(database.read_people(
+            session=act_session, full_name=_FULL_NAME, birthdate=_BIRTHDATE))
+        queried_person.bookings[0].admission_date = booking_admission_date
```
Comment (terinpw) [inline recidiviz/tests/persistence/database/database_test.py]: nit use `one` here?

Comment (zdg2102) [reply recidiviz/tests/persistence/database/database_test.py]: I think for situations like this where I have to access something down the record tree to change something, using the index is more readable than having `one` calls

**Pull request merged**